File: RaWin/ViewModels/FeaturesExplorerPanelViewModel.cs
using System;
using System.Collections.ObjectModel;
using System.Text.RegularExpressions;
using System.Windows.Input;
using RaCore.Engine.Manager;

namespace RaWin.ViewModels
{
    public class FeatureExplorerPanelViewModel : ObservableObject
    {
        private readonly IRaModule? _featureExplorer;

        public string FeatureReport { get; private set; } = "(No report yet)";
        public ObservableCollection<string> Modules { get; } = new();

        private string _errorMessage = "";
        public string ErrorMessage
        {
            get => _errorMessage;
            set { _errorMessage = value; OnPropertyChanged(); }
        }

        public ICommand ReloadCommand { get; }

        public FeatureExplorerPanelViewModel(ModuleManager manager)
        {
            _featureExplorer = manager.GetModuleByName("FeatureExplorer");
            ReloadCommand = new RelayCommand(_ => LoadFeatureReport());
            LoadFeatureReport();
        }

        public void LoadFeatureReport()
        {
            try
            {
                Modules.Clear();
                if (_featureExplorer != null)
                {
                    FeatureReport = _featureExplorer.Process("features full") ?? "(No report received)";
                    var lines = FeatureReport.Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries);
                    var moduleLineRegex = new Regex(@"^\s*-\s*(\w+)\s*\(");
                    foreach (var line in lines)
                    {
                        var match = moduleLineRegex.Match(line);
                        if (match.Success)
                        {
                            Modules.Add(match.Groups[1].Value);
                        }
                    }
                    ErrorMessage = "";
                }
                else
                {
                    FeatureReport = "(FeatureExplorer module not loaded)";
                    ErrorMessage = "FeatureExplorer module not loaded";
                }
            }
            catch (Exception ex)
            {
                FeatureReport = "(Error loading report)";
                ErrorMessage = $"Error: {ex.GetType().Name}: {ex.Message}";
            }
            OnPropertyChanged(nameof(FeatureReport));
            OnPropertyChanged(nameof(Modules));
            OnPropertyChanged(nameof(ErrorMessage));
        }
    }
}

File: RaStudios/viewmodels/features_explorer_panel_viewmodel.py
from viewmodels.observable_object import ObservableObject
from viewmodels.relay_command import RelayCommand
import re

class FeatureExplorerPanelViewModel(ObservableObject):
    """
    ViewModel for feature/module explorer panel.
    Supports plugin features, manifest metadata, and diagnostics via SpeechPipelineService.
    """
    def __init__(self, speech_pipeline, module_bus):
        super().__init__()
        self.speech_pipeline = speech_pipeline
        self.module_bus = module_bus
        self.feature_report = "(No report yet)"
        self.modules = []
        self.error_message = ""
        self.reload_command = RelayCommand(self.load_feature_report)
        self.load_feature_report()

    def load_feature_report(self, _=None):
        try:
            self.modules.clear()
            self.feature_report = self.speech_pipeline.send_async("features full")
            lines = self.feature_report.splitlines()
            module_line_regex = re.compile(r"^\s*Module:\s*(\w+)")
            for line in lines:
                match = module_line_regex.match(line)
                if match:
                    self.modules.append(match.group(1))
            self.error_message = ""
        except Exception as ex:
            self.feature_report = "(Error loading report)"
            self.error_message = f"Error: {type(ex).__name__}: {ex}"
            self.module_bus.publish("Diagnostics", {
                "timestamp": "now",
                "message": self.error_message,
                "severity": "Error",
                "source": "FeatureExplorerPanelViewModel",
                "code": "FEATURE_REPORT_ERROR",
                "metadata": {"FeatureReport": self.feature_report}
            })
        self.notify_property_changed("feature_report")
        self.notify_property_changed("modules")
        self.notify_property_changed("error_message")